File: deploy/local-preview/preview-server.py
#!/usr/bin/env python3
"""
RS Chef'z - local preview server (Python version).

Identical in behaviour to preview-server.mjs; use whichever of Python or
Node you already have. Serves the folder it sits in the way Hostinger
will, so the preview matches the live site.

It does the four things opening the file directly cannot:

  1. Acts as a real web server, so "/_next/..." resolves to this folder
     rather than the top of your hard disk.
  2. Implements the .htaccess clean-URL rule, so
     /products/gobi-manchurian-masala works without the .html.
  3. Answers byte-range requests, which <video> requires.
  4. Listens on the network and prints the address to open on a phone or
     tablet over the same Wi-Fi.

Run:  python preview-server.py
"""

import http.server
import os
import posixpath
import re
import socket
import sys
import threading
import urllib.parse
import webbrowser

HERE = os.path.dirname(os.path.abspath(__file__))

# The download keeps the site in its own folder, so that folder is exactly
# what gets uploaded and nothing else has to be deleted first. Serve it if it
# is there; otherwise serve whatever folder this file is sitting in, which is
# what happens once the launcher has been copied in beside index.html.
SITE_FOLDER = "UPLOAD-TO-public_html"
ROOT = (
    os.path.join(HERE, SITE_FOLDER)
    if os.path.isfile(os.path.join(HERE, SITE_FOLDER, "index.html"))
    else HERE
)

PORT = int(os.environ.get("PORT", "8080"))

TYPES = {
    ".html": "text/html; charset=utf-8",
    ".js": "text/javascript; charset=utf-8",
    ".mjs": "text/javascript; charset=utf-8",
    ".css": "text/css; charset=utf-8",
    ".json": "application/json; charset=utf-8",
    ".txt": "text/plain; charset=utf-8",
    ".xml": "application/xml; charset=utf-8",
    ".webp": "image/webp",
    ".avif": "image/avif",
    ".jpg": "image/jpeg",
    ".jpeg": "image/jpeg",
    ".png": "image/png",
    ".gif": "image/gif",
    ".svg": "image/svg+xml",
    ".ico": "image/x-icon",
    ".mp4": "video/mp4",
    ".webm": "video/webm",
    ".woff2": "font/woff2",
    ".woff": "font/woff",
}

RANGE_RE = re.compile(r"^bytes=(\d*)-(\d*)$")


class Handler(http.server.BaseHTTPRequestHandler):
    protocol_version = "HTTP/1.1"

    def log_message(self, fmt, *args):
        pass  # quiet

    def resolve(self, path):
        """Map a URL path to a file, applying the clean-URL rule."""
        path = urllib.parse.unquote(path.split("?", 1)[0].split("#", 1)[0])
        path = posixpath.normpath(path)
        parts = [p for p in path.split("/") if p and p not in (".", "..")]
        target = os.path.join(ROOT, *parts)

        if path == "/" or not parts:
            return os.path.join(ROOT, "index.html")
        if os.path.isfile(target):
            return target
        if os.path.isfile(os.path.join(target, "index.html")):
            return os.path.join(target, "index.html")
        if not os.path.splitext(target)[1] and os.path.isfile(target + ".html"):
            return target + ".html"
        return None

    def do_HEAD(self):
        self.respond(head_only=True)

    def do_GET(self):
        self.respond()

    def respond(self, head_only=False):
        raw = self.path.split("?", 1)[0]
        if len(raw) > 1 and raw.endswith("/"):
            self.send_response(301)
            self.send_header("Location", raw[:-1])
            self.send_header("Content-Length", "0")
            self.end_headers()
            return

        target = self.resolve(self.path)

        if target is None:
            nf = os.path.join(ROOT, "404.html")
            body = open(nf, "rb").read() if os.path.isfile(nf) else b"404 Not Found"
            self.send_response(404)
            self.send_header("Content-Type", "text/html; charset=utf-8")
            self.send_header("Content-Length", str(len(body)))
            self.end_headers()
            if not head_only:
                self.wfile.write(body)
            return

        with open(target, "rb") as fh:
            body = fh.read()
        ctype = TYPES.get(os.path.splitext(target)[1].lower(), "application/octet-stream")

        rng = self.headers.get("Range")
        if rng:
            m = RANGE_RE.match(rng.strip())
            if m:
                size = len(body)
                start = int(m.group(1)) if m.group(1) else 0
                end = int(m.group(2)) if m.group(2) else size - 1
                if start <= end < size:
                    chunk = body[start : end + 1]
                    self.send_response(206)
                    self.send_header("Content-Type", ctype)
                    self.send_header("Content-Range", f"bytes {start}-{end}/{size}")
                    self.send_header("Accept-Ranges", "bytes")
                    self.send_header("Content-Length", str(len(chunk)))
                    self.send_header("Cache-Control", "no-store")
                    self.end_headers()
                    if not head_only:
                        self.wfile.write(chunk)
                    return

        self.send_response(200)
        self.send_header("Content-Type", ctype)
        self.send_header("Accept-Ranges", "bytes")
        self.send_header("Content-Length", str(len(body)))
        self.send_header("Cache-Control", "no-store")
        self.end_headers()
        if not head_only:
            self.wfile.write(body)


def lan_addresses():
    found = []
    try:
        s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        s.connect(("8.8.8.8", 80))  # no packet sent; just picks the route
        found.append(s.getsockname()[0])
        s.close()
    except Exception:
        pass
    return found


if __name__ == "__main__":
    try:
        server = http.server.ThreadingHTTPServer(("0.0.0.0", PORT), Handler)
    except OSError as err:
        print(f"\n  Could not start on port {PORT}: {err}")
        print("  Close the other preview window, or pick another port:")
        print("      PORT=8081 python preview-server.py\n")
        sys.exit(1)

    print("\n  RS Chef'z preview is running.\n")
    print(f"  On this computer:   http://localhost:{PORT}")
    lan = lan_addresses()
    if lan:
        print("\n  On your phone or tablet (same Wi-Fi):")
        for ip in lan:
            print(f"                      http://{ip}:{PORT}")
    print("\n  Leave this window open. Press Ctrl+C to stop.\n")

    # Open the browser, so the whole thing is one double-click and nobody
    # has to type an address anywhere.
    if os.environ.get("NO_OPEN") != "1":
        threading.Timer(
            0.4, lambda: webbrowser.open(f"http://localhost:{PORT}")
        ).start()

    try:
        server.serve_forever()
    except KeyboardInterrupt:
        print("\n  Preview stopped.\n")

File: deploy/local-preview/preview-server.mjs
/**
 * RS Chef'z — local preview server (Node version).
 *
 * Serves the folder it is sitting in the way Hostinger will, so what you
 * see locally is what goes live. It does the four things a plain file
 * open cannot:
 *
 *   1. Acts as a real web server, so "/_next/..." resolves to this folder
 *      instead of the top of your hard disk. That alone is the difference
 *      between a blank serif page and the site.
 *   2. Implements the .htaccess clean-URL rule, so
 *      /products/gobi-manchurian-masala works without the .html.
 *   3. Answers byte-range requests, which <video> requires. Without them
 *      the browser refuses the film even though the file is perfect.
 *   4. Listens on every network interface and prints the address your
 *      phone and tablet can use over the same Wi-Fi.
 *
 * Run:  node preview-server.mjs
 */
import { createServer } from "node:http";
import { readFile, stat } from "node:fs/promises";
import { existsSync } from "node:fs";
import { join, extname, dirname, normalize } from "node:path";
import { fileURLToPath } from "node:url";
import { networkInterfaces, platform } from "node:os";
import { spawn } from "node:child_process";

const HERE = dirname(fileURLToPath(import.meta.url));

/*
 * The download keeps the site in its own folder, so that folder is exactly
 * what gets uploaded and nothing else has to be deleted first. Serve it if
 * it is there; otherwise serve whatever folder this file is sitting in,
 * which is what happens once the launcher has been copied in beside
 * index.html.
 */
const SITE_FOLDER = "UPLOAD-TO-public_html";
const ROOT = existsSync(join(HERE, SITE_FOLDER, "index.html"))
  ? join(HERE, SITE_FOLDER)
  : HERE;

const PORT = Number(process.env.PORT ?? 8080);

/*
 * The permanent redirects .htaccess serves, repeated here so the preview
 * tells the same truth as Hostinger. Without them an old address 404s
 * locally and silently works live, which is the wrong way round for a thing
 * whose job is to show you what you are about to upload.
 */
const REDIRECTS = new Map([
  ["/products/three-in-one-masala", "/products/chicken-65-masala"],
]);

const TYPES = {
  ".html": "text/html; charset=utf-8",
  ".js": "text/javascript; charset=utf-8",
  ".mjs": "text/javascript; charset=utf-8",
  ".css": "text/css; charset=utf-8",
  ".json": "application/json; charset=utf-8",
  ".txt": "text/plain; charset=utf-8",
  ".xml": "application/xml; charset=utf-8",
  ".webp": "image/webp",
  ".avif": "image/avif",
  ".jpg": "image/jpeg",
  ".jpeg": "image/jpeg",
  ".png": "image/png",
  ".gif": "image/gif",
  ".svg": "image/svg+xml",
  ".ico": "image/x-icon",
  ".mp4": "video/mp4",
  ".webm": "video/webm",
  ".woff2": "font/woff2",
  ".woff": "font/woff",
};

async function isFile(p) {
  try {
    return (await stat(p)).isFile();
  } catch {
    return false;
  }
}

function lanAddresses() {
  const out = [];
  for (const list of Object.values(networkInterfaces())) {
    for (const net of list ?? []) {
      if (net.family === "IPv4" && !net.internal) out.push(net.address);
    }
  }
  return out;
}

const server = createServer(async (req, res) => {
  try {
    const url = new URL(req.url, "http://localhost");
    let path = decodeURIComponent(url.pathname);

    // Never let a request climb out of this folder.
    path = normalize(path).replace(/^(\.\.[/\\])+/, "");

    if (path.length > 1 && path.endsWith("/")) {
      res.writeHead(301, { Location: path.slice(0, -1) + url.search });
      return res.end();
    }

    const moved = REDIRECTS.get(path);
    if (moved) {
      res.writeHead(301, { Location: moved + url.search });
      return res.end();
    }

    let file = path === "/" ? join(ROOT, "index.html") : join(ROOT, path);

    if (!(await isFile(file))) {
      if (await isFile(join(file, "index.html"))) {
        file = join(file, "index.html");
      } else if (!extname(path) && (await isFile(`${file}.html`))) {
        // The .htaccess clean-URL rule.
        file = `${file}.html`;
      } else {
        const notFound = join(ROOT, "404.html");
        const body = (await isFile(notFound))
          ? await readFile(notFound)
          : "404 Not Found";
        res.writeHead(404, { "Content-Type": "text/html; charset=utf-8" });
        return res.end(body);
      }
    }

    const body = await readFile(file);
    const type = TYPES[extname(file).toLowerCase()] ?? "application/octet-stream";

    // Byte ranges, for <video>.
    const range = req.headers.range;
    if (range) {
      const m = /^bytes=(\d*)-(\d*)$/.exec(range);
      if (m) {
        const size = body.length;
        const start = m[1] ? Number(m[1]) : 0;
        const end = m[2] ? Number(m[2]) : size - 1;
        if (start <= end && end < size) {
          res.writeHead(206, {
            "Content-Type": type,
            "Content-Range": `bytes ${start}-${end}/${size}`,
            "Accept-Ranges": "bytes",
            "Content-Length": end - start + 1,
            "Cache-Control": "no-store",
          });
          return res.end(body.subarray(start, end + 1));
        }
      }
    }

    res.writeHead(200, {
      "Content-Type": type,
      "Accept-Ranges": "bytes",
      "Content-Length": body.length,
      // Never cache locally, so a rebuild always shows.
      "Cache-Control": "no-store",
    });
    res.end(body);
  } catch {
    res.writeHead(500, { "Content-Type": "text/plain" });
    res.end("500");
  }
});

server.on("error", (err) => {
  if (err.code === "EADDRINUSE") {
    console.error(
      `\n  Port ${PORT} is already being used by something else.` +
        `\n  Close the other preview window, or start this one on a different` +
        `\n  port:   PORT=8081 node preview-server.mjs\n`,
    );
    process.exit(1);
  }
  throw err;
});

/**
 * Open the site in the default browser, so the whole thing is one
 * double-click and nobody has to type an address anywhere.
 */
function openBrowser(url) {
  const cmd =
    platform() === "win32"
      ? ["cmd", ["/c", "start", "", url]]
      : platform() === "darwin"
        ? ["open", [url]]
        : ["xdg-open", [url]];
  try {
    const child = spawn(cmd[0], cmd[1], { stdio: "ignore", detached: true });
    /*
     * A missing opener is reported asynchronously, as an 'error' event —
     * spawn itself does not throw, so the try/catch around it never sees
     * one. An 'error' event with nothing listening is thrown at the process
     * instead, which would take the whole server down: the window closes
     * the instant it opens and the preview looks broken when it is not.
     * Swallow it. Not being able to launch a browser is not a reason to
     * stop serving; the address is printed just below.
     */
    child.on("error", () => {});
    child.unref();
  } catch {
    /* No browser to launch is not a reason to stop serving. */
  }
}

server.listen(PORT, "0.0.0.0", () => {
  console.log(`\n  RS Chef'z preview is running.\n`);
  console.log(`  On this computer:   http://localhost:${PORT}`);
  const lan = lanAddresses();
  if (lan.length) {
    console.log(`\n  On your phone or tablet (same Wi-Fi):`);
    for (const ip of lan) console.log(`                      http://${ip}:${PORT}`);
  } else {
    console.log(`\n  No network address found — phone access needs Wi-Fi.`);
  }
  console.log(`\n  Leave this window open. Press Ctrl+C to stop.\n`);
  if (process.env.NO_OPEN !== "1") openBrowser(`http://localhost:${PORT}`);
});
